File: 05-binary_search/00981-Time_Based_Key-Value_Store/00981-time_based_key-value_store.cpp
// 981: Time Based Key-Value Store
// https://leetcode.com/problems/time-based-key-value-store/


#include <iostream>
#include <unordered_map>
#include <vector>
#include <string>

using namespace std;


// SOLUTION
class TimeMap {
private:
    unordered_map<string, vector<pair<int, string>>> tm;

public:
    TimeMap() {}

    void set(string key, string value, int timestamp) {
        tm[key].push_back({timestamp, value});
    }
    
    string get(string key, int timestamp) {
        if (tm.find(key) == tm.end()) return "";

        int l = 0;
        int h = tm[key].size() - 1;

        while (l <= h) {
            int m = l + (h - l) / 2;

            if (tm[key][m].first < timestamp)
                l = m + 1;
            else if (tm[key][m].first > timestamp)
                h = m - 1;
            else
                return tm[key][m].second;
        }
        
        if (h >= 0) return tm[key][h].second;

        return "";
    }
};


int main() {
    // OPERATIONS
    TimeMap* o = new TimeMap();
    o->set("foo", "bar", 1);
    cout << o->get("foo", 1) << endl;
    cout << o->get("foo", 3) << endl;
    o->set("foo", "bar2", 4);
    cout << o->get("foo", 4) << endl;
    cout << o->get("foo", 5) << endl;

    return 0;
}

File: 05-binary_search/00981-Time_Based_Key-Value_Store/00981-time_based_key-value_store.go
// 981: Time Based Key-Value Store
// https://leetcode.com/problems/time-based-key-value-store/

package main

import (
	"fmt"
	"sort"
)

type Data struct {
    val string
    time int
}

type TimeMap struct {
    tm map[string][]Data
}


func Constructor() TimeMap {
    return TimeMap{tm: make(map[string][]Data)}
}


func (this *TimeMap) Set(key string, value string, timestamp int)  {
    this.tm[key] = append(
        this.tm[key], 
        Data{time: timestamp,val: value},
    )
}


func (this *TimeMap) Get(key string, timestamp int) string {
    var list []Data
    var ok bool
    
    if list, ok = this.tm[key]; !ok {
        return ""
    }

    index := sort.Search(len(list), func(i int) bool {
        return list[i].time > timestamp
    })
    
    if index == 0 {
        return ""
    } else {
        return list[index - 1].val
    }  
}

func main() {
    // OPERATIONS
    o := Constructor()
    o.Set("foo", "bar", 1)
    fmt.Println(o.Get("foo", 1))
    fmt.Println(o.Get("foo", 3))
    o.Set("foo", "bar2", 4)
    fmt.Println(o.Get("foo", 4))
    fmt.Println(o.Get("foo", 5))
}